lib/Parse/ParseStmt.cpp

```diff
@@ -901,6 +908,99 @@ ParserResult<Stmt> Parser::parseStmtYield(SourceLoc tryLoc) {
            status, YieldStmt::create(Context, yieldLoc, lpLoc, yields, rpLoc));
 }
 
+bool Parser::isContextualThenKeyword(bool preferExpr) {
+  if (!Context.LangOpts.hasFeature(Feature::ThenStatements))
+    return false;
+
+  if (!Tok.isContextualKeyword("then"))
+    return false;
+
+  // If we want to prefer an expr, and aren't at the start of a newline, then
+  // don't parse a ThenStmt.
+  if (preferExpr && !Tok.isAtStartOfLine())
+    return false;
+
+  // 'then' immediately followed by '('/'[' is a function/subscript call. If
+  // immediately followed by '.', it's a member access.
+  //
+  // Note we handle the '.' case before the operator case, we want to maintain
+  // compatibility with the new SwiftSyntax parser, which will consider
+  // 'then . foo' to be a 'then' statement, not a binary operator.
+  if (peekToken().isAny(tok::l_paren, tok::l_square, tok::period)) {
+    auto nextLoc = Lexer::getLocForEndOfToken(SourceMgr, Tok.getLoc());
```

Comment:
I’d name this variable `tokEndLoc`. `nextLoc` sounds like the location of `peekToken()`.